### ISSUE TITLE
Native support for AssumeRole flow in NetCore.Setup

### ISSUE THREAD

**Issue by Kralizek** [aws/aws-sdk-net#1741]
It would be extremely useful if the AWS SDK supported natively the possibility for assuming a role. 

This is explicitly useful in development phase because it would allow
- IAM/SSO users with very limited permissions (ideally only the possibility to assume roles)
- applications would run also in development with their assigned role thus making it easier to test the role itself before going into production

## Describe the Feature
Starting from this page of the documentation: https://docs.aws.amazon.com/sdk-for-net/v3/developer-guide/net-dg-config-netcore.html

It would be cool if we could have native support for something like 

```json
// application.Development.json
{
	"AWS": {
		"Region": "eu-north-1",
		"Role": "my-role-to-assume"
	}
}
```

Specifying a role in the configuration should change the workflow for instantiating a service client (e.g. s3) from "fetch credentials from shared file" to
- fetch credentials to shared file
- instantiate sts client with these credentials
- assume specified role
- fetch new credentials
- instantiate service client with new credentials

### Is your Feature Request related to a problem?
In a way it can be done already but I believe it should be part of the SDK and not part of the application code because
- this is purely devops plumbing
- the setup is expected to change between development and production environment and I'd avoid ugly ifs as much as possible
- during development, the applications are running with the users credentials that are usually granted many permissions. This creates friction when moving into production where usually roles are being used.

## Proposed Solution
When a key AWS:Role (or equivalent according to customizability of the setup) is present, the SDK should resolve service client interfaces (e.g. `IAmazonS3`) using the STS client to assume the role and fetch temporary credentials.

Most of the work would be done [here](https://github.com/aws/aws-sdk-net/blob/e9603c9005f6b43054c8bc5e1f826189c99b5395/extensions/src/AWSSDK.Extensions.NETCore.Setup/ClientFactory.cs#L125-L162).

### Describe alternatives you've considered
Polluting the .aws\credentials file with many smaller profiles.

## Additional Context
PROs:
- betters security because IAM users can have less permissions
- better development flow because applications will be developed with their permissions from the beginning

CONs:
- NetCore.Setup would need to take a dependency on the SecurityToken package.

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change  

I might be able to put a simple PR but there are some corner cases that should be investigated properly (e.g. cross account AssumeRole).

---

This is a :rocket: Feature Request

**Comment by McDoit** [comment]
This would be a very nice feature for us as well!

**Comment by Kralizek** [comment]
Hi!

I would love to receive any feedback of some sort regarding this issue and, if needed, the connected pull request.

Thanks!

**Comment by dferretti** [comment]
This would be very useful for us too! Our use case is when running in ECS and we want to reach out to resources in another account. The ECS task has the correct AssumeRole permissions, but I need a quick way for telling the c# SDK to actually use that role. Similar to above, we are also polluting the project with `config` and `credentials` files with a dummy role pointing to the ARN we want to use, and using the dummy role in our AWSOptions. Cutting out the middleman and letting us configure the SDK with just a role arn would be preferable.
To me, I see profiles as a convenience for a developer machine that needs friendly-named roles to switch between for CLI or other purposes. But for deploying, the profile name is just an extra hop between us and the role ARN we already had to somehow configure anyway.

**Comment by ashishdhingra** [comment]
Needs to be reviewed by the team.

**Comment by normj** [comment]
PR has been released as part of version [3.7.3](https://www.nuget.org/packages/AWSSDK.Extensions.NETCore.Setup/3.7.3) adding the new SessionRoleArn and SessionName properties.
